File: cpp/results.py
import sys
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
from parse import *

read_time = 10 * 10**-9
write_time = 100 * 10**-9
adc_latency = 1 * 10**-9
read_energy = 40 * 10**-15
write_energy = 20 * 10**-12
adc_energy = 2 * 10**-12
sa_latency = 1 * 10**-9
sa_energy = 0.01 * 10**-12

#experiments = ["wiki-Vote", "amazon0302", "com-lj", "web-Google",
 #       "soc-Slashdot0902.txt"]
#experiments = ["wiki-Vote", "web-Google", "soc-Slashdot0902", "amazon0302"]
experiments = ["wiki-Vote", "soc-Slashdot0902", "amazon0302", "soc-Epinions1"]

def values_from_file(filename, experiment):
    print("doing experiment: " + filename)
    with open(filename, "r") as results:
        data = results.readlines()
        total_time = 0
        total_energy = 0
        efficiency = 0.0
        total_periphery_time = 0
        total_periphery_energy = 0
        [row, col, sense_type] = parse("init: {:d},{:d},{:l}\n", data[0]) 

        energy = 0
        latency = 0

        if sense_type == "adc":
            energy = adc_energy
            latency = adc_latency
        elif sense_type == "sa":
            energy = sa_energy
            latency = sa_latency
        else:
            print("Unknown sense_type: " + sense_type)

        dynamic_latency = 0
        dynamic_energy = 0
        static_energy = 0
        static_time = 0
        if experiment == "ours":
            dynamic_latency = 1.1 * 10 **-9
            dynamic_energy = 28.7 * 10 ** -12

        if experiment == "graphr":
            static_energy = 11.8 * 10 ** -12
            static_time = 0.5 * 10 ** -9

        for i in range(1, len(data)):
            parsed = parse("{:l}{}", data[i])
            command = parsed[0].strip('\n')
            if command == "clear":
                total_time += row*write_time
                total_energy += row*write_energy
            elif command == "read":
                [adc_acts, row_reads, row_writes, inputs] = parse(" {:d},{:d},{:d},{:d}\n", parsed[1])
                total_time += (read_time + adc_acts * latency) * inputs
                total_energy += (row_reads * read_energy + adc_acts * energy) * inputs
                if experiment == "graphr":
                    total_periphery_energy += static_energy * col
                    total_periphery_time += static_time
            elif command == "write":
                [adc_acts, row_reads, row_writes, inputs] = parse(" {:d},{:d},{:d},{:d}\n", parsed[1])
                total_time += (write_time + adc_acts * latency) * inputs
                total_energy += (row_writes * write_energy + adc_acts * energy) * inputs
            elif command == "efficiency":
                efficiency = parse(" {:f}\n", parsed[1])[0]
            elif command == "elements":
                elements = parse(" {:d}\n", parsed[1])[0]
                total_periphery_time += elements * dynamic_latency
                total_periphery_energy += elements * dynamic_energy
            else:
                print("Unknown command " + command)
        return (total_time, total_energy, efficiency, total_periphery_time, total_periphery_energy)


experiment_results = {
        "graphr-time": [],
        "graphr-energy": [],
        "graphr-efficiency": [],
        "our-time": [],
        "our-energy": [],
        "our-efficiency": [],
        "our-offsets-time" : [],
        "our-offsets-energy": [],
        "our-total-time": [],
        "our-total-energy": [],
        "normal-graphr-time": [],
        "normal-graphr-energy": [],
        "normal-graphr-efficiency": []}

# for experiment in experiments:
#     [graphr_time, graphr_energy, graphr_efficiency, graphr_periphery_time, graphr_periphery_energy] = values_from_file(experiment + "-graphr.log", "graphr")
#     experiment_results["graphr-time"].append(graphr_time)
#     experiment_results["graphr-energy"].append(graphr_energy)
#     experiment_results["graphr-efficiency"].append(graphr_efficiency)
# 
#     [our_time, our_energy, our_efficiency, our_periphery_time, our_periphery_energy] = values_from_file(experiment + "-ours.log", "ours")
#     experiment_results["our-time"].append(our_time)
#     experiment_results["our-energy"].append(our_energy)
#     experiment_results["our-efficiency"].append(our_efficiency)
# 
#     [our_offset_time, our_offset_energy, _, _, _] = values_from_file(experiment + "-ours-offsets.log", "offsets")
#     experiment_results["our-offsets-time"].append(our_offset_time)
#     experiment_results["our-offsets-energy"].append(our_offset_energy)
# 
#     our_total_energy = our_offset_energy + our_energy + our_periphery_energy
#     our_total_time = our_offset_time + our_time + our_periphery_time
# 
#     graphr_total_energy = graphr_energy + graphr_periphery_energy
#     graphr_total_time = graphr_time + graphr_periphery_time
# 
#     experiment_results["our-total-time"].append(our_offset_time + our_time)
#     experiment_results["our-total-energy"].append(our_offset_energy + our_energy)
# 
#     experiment_results["normal-graphr-time"].append(graphr_total_time / our_total_time)
#     experiment_results["normal-graphr-energy"].append(graphr_total_energy / our_total_energy)
#     experiment_results["normal-graphr-efficiency"].append(graphr_efficiency / our_efficiency)

# Wiki-Vote
experiment_results["normal-graphr-time"].append(102.234)
experiment_results["normal-graphr-energy"].append(458.843)
experiment_results["normal-graphr-efficiency"].append(0.895387)

experiment_results["normal-graphr-time"].append(157.172)
experiment_results["normal-graphr-energy"].append(790.718)
experiment_results["normal-graphr-efficiency"].append(0.307994)

experiment_results["normal-graphr-time"].append(31.7518)
experiment_results["normal-graphr-energy"].append(205.106)
experiment_results["normal-graphr-efficiency"].append(0.126417)

experiment_results["normal-graphr-time"].append(156.015)
experiment_results["normal-graphr-energy"].append(1102.22)
experiment_results["normal-graphr-efficiency"].append(0.173824)

print(experiment_results)

colors = sns.color_palette("pastel", as_cmap=True)
graph_colors = {
    'our': colors[0],
    'graphr': colors[1]
}

width = 0.25
x = np.arange(len(experiments))
normals = [1] * len(experiments)

fig, ax = plt.subplots()
rects1 = ax.bar(x-width/2, experiment_results["normal-graphr-time"], width,
        label='GraphR', color=graph_colors['graphr'])
rects2 = ax.bar(x+width/2, normals, width,
        label='Our', color=graph_colors['our'])
ax.legend()
ax.set_yscale('log')
ax.set_xticks(x, experiments)
ax.bar_label(rects1, padding=3)
ax.bar_label(rects2, padding=3)
ax.set_xlabel("Dataset")
ax.set_ylabel("Time")
ax.set_title("Normalized time difference per dataset")

fig.tight_layout()
plt.savefig('time.png')

fig, ax = plt.subplots()
rects1 = ax.bar(x-width/2, experiment_results["normal-graphr-energy"], width,
        label='GraphR', color=graph_colors['graphr'])
rects2 = ax.bar(x+width/2, normals, width,
        label='Our', color=graph_colors['our'])

ax.legend()
ax.set_yscale('log')
ax.set_xticks(x, experiments)
ax.bar_label(rects1, padding=3)
ax.bar_label(rects2, padding=3)
ax.set_xlabel("Dataset")
ax.set_ylabel("Energy")
ax.set_title("Normalized energy difference per dataset")

fig.tight_layout()
plt.savefig('energy.png')

fig, ax = plt.subplots()
rects1 = ax.bar(x-width/2, experiment_results["normal-graphr-efficiency"], width,
        label='GraphR', color=graph_colors['graphr'])
rects2 = ax.bar(x+width/2, normals, width,
        label='Our', color=graph_colors['our'])

ax.legend()
ax.set_xticks(x, experiments)
ax.bar_label(rects1, padding=3)
ax.bar_label(rects2, padding=3)
ax.set_xlabel("Dataset")
ax.set_ylabel("Efficiency")
ax.set_title("Normalized efficiency difference per dataset")

fig.tight_layout()
plt.savefig('efficiency.png')

File: cpp/results.cpp
#include <iostream>
#include <stdio.h>
#include <string>
#include <string.h>

const char *experiments[] = {"wiki-Vote", "soc-Slashdot0902", "amazon0302", "soc-Epinions1"};

struct Result {
	float total_time = 0, total_energy = 0, efficiency = 0,
	      total_periphery_time = 0, total_periphery_energy = 0;
};

static constexpr float READ_TIME = 10e-9;
static constexpr float WRITE_TIME = 100e-9;
static constexpr float ADC_LATENCY = 1e-9;
static constexpr float READ_ENERGY = 40e-15;
static constexpr float WRITE_ENERGY = 20e-12;
static constexpr float ADC_ENERGY = 2e-12;
static constexpr float SA_LATENCY = 1e-9;
static constexpr float SA_ENERGY = 10e-12;

static Result values_from_file(const std::string &filename, const std::string &experiment) {
	std::cout << "doing experiment " << filename << std::endl;
	Result res;
	FILE *fp = fopen(filename.c_str(), "r");

	float static_energy = 0, static_time = 0;
	if (experiment == "graphr") {
		static_energy = 11.8e-12;
		static_time = 0.5e-9;
	}

	float dynamic_energy = 0, dynamic_latency = 0;
	if (experiment == "ours") {
		dynamic_latency = 1.1e-9;
		dynamic_energy = 28.8e-12;
	}

	float energy = 0, latency = 0;

	char line[1024];
	char sense_type[128];
	size_t row, col;
	while (fgets(line, sizeof(line), fp)) {
		char command[128];
		int read;
		sscanf(line, "%s %n", command, &read);
		if (strcmp(command, "init:") == 0) {
			sscanf(line + read, "%ld,%ld,%s\n", &row, &col, sense_type);
			if (strcmp(sense_type, "adc") == 0) {
				energy = ADC_ENERGY;
				latency = ADC_LATENCY;
			} else if (strcmp(sense_type, "sa") == 0) {
				energy = SA_ENERGY;
				latency = SA_LATENCY;
			} else {
				std::cout << "Unkown sense type: " << sense_type << std::endl;
			}
		} else if (strcmp(command, "clear") == 0) {
			res.total_time += row * WRITE_TIME;
			res.total_energy += row * WRITE_ENERGY;
		} else if (strcmp(command, "read") == 0) {
			int adc_acts, row_reads, row_writes, inputs;
			sscanf(line + read, "%d,%d,%d,%d\n", &adc_acts, &row_reads, &row_writes,&inputs);
			res.total_time += (READ_TIME + adc_acts * latency) * inputs;
			res.total_energy += (row_reads * READ_ENERGY + adc_acts * energy) * inputs;
			if (experiment == "graphr") {
				res.total_periphery_energy += static_energy * col;
				res.total_periphery_time += static_time;
			}
		} else if (strcmp(command, "write") == 0) {
			int adc_acts, row_reads, row_writes, inputs;
			sscanf(line + read, "%d,%d,%d,%d\n", &adc_acts, &row_reads, &row_writes,&inputs);
			res.total_time += (WRITE_TIME) * inputs;
			res.total_energy += (row_writes * WRITE_ENERGY) * inputs;
		} else if (strcmp(command, "efficiency") == 0) {
			float efficiency;
			sscanf(line + read, "%f\n", &efficiency);
			res.efficiency = efficiency;
		} else if (strcmp(command, "elements") == 0) {
			int elements;
			sscanf(line + read, "%d\n", &elements);
			res.total_periphery_time += elements * dynamic_latency;
			res.total_periphery_energy += elements * dynamic_energy;
		} else {
			std::cout << "unkown command " << command << std::endl;
			exit(1);
		}
	}

	fclose(fp);
	return res;
}

int main() {
	for (auto experiment : experiments) {
		auto experiment_str = std::string(experiment);

		auto graphr_res = values_from_file(experiment_str + "-graphr.log", "graphr");
		auto our_res = values_from_file(experiment_str + "-ours.log", "ours");
		auto our_offsets_res = values_from_file(experiment_str + "-ours-offsets.log", "offsets");

		auto our_total_energy = our_res.total_energy + our_offsets_res.total_energy
			+ our_res.total_periphery_energy;
		auto our_total_time = our_res.total_time + our_offsets_res.total_time
			+ our_res.total_periphery_time;

		auto graphr_total_energy = graphr_res.total_energy
			+ graphr_res.total_periphery_energy;
		auto graphr_total_time = graphr_res.total_time
			+ graphr_res.total_periphery_time;

		std::cout << "normal graphr time: " << graphr_total_time / our_total_time << std::endl;
		std::cout << "normal graphr energy: " << graphr_total_energy / our_total_energy << std::endl;
		std::cout << "normal graphr efficiency: " << graphr_res.efficiency / our_res.efficiency << std::endl;
	}
	return 0;
}
